cilium/CFP-32810-hybrid-routing-mode.md

```diff
@@ -0,0 +1,95 @@
+
+With higher adoption in cluster mesh and large scale clusters being common to support AI workloads, it would be beneficial to have a more performant dataplane that can optimize the throughput, while providing support for both encapsulation and native-routing.
+
+## Goals
+
+* Retain the behavior of current routing modes as-is
+* Allow users to set a new routing mode called `hybrid`
+* Allow users to pass in network topologies expressed as CIDR ranges through configmap
+* Support for both IPV4 and IPV6
+* Allow dynamic configuration of CIDRs (add/remove) without agent restart (update w/o agent restart)
+
+## Proposal
+
+### Overview
+
+Add a new option for routing-mode config as follows `routing-mode: hybrid`. Also, add a new field in config to pass information about the CIDRs and connectivity. The CIDR information will persist in pinned eBPF maps so that routing decision can be made in runtime by eBPF programs. The concept is similar to [IP Masquerading](https://docs.cilium.io/en/stable/network/concepts/masquerading/#ebpf-based) and the eBPF table it uses.
+
+### Examples
+
+```
+subnet-topology: 10.0.0.1/24
+```
+
+If source and destination IP belongs to above CIDR, route natively. Else, preserve current behavior (encapsulate if being routed to known CIDRs/pass to Linux stack).
+
+```
+subnet-topology: 10.0.0.1/24,10.10.0.1/24
+```
+
+The above two CIDR ranges have direct connectivity. If source and destination IP belongs to either CIDR, route natively. Else, preserve current behavior (encapsulate if being routed to know CIDRs/pass to Linux stack).
+```
+subnet-topology: 10.0.0.1/24,10.10.0.1/24;10.20.0.1/24
+```
+
+* `10.0.0.1/24` and `10.10.0.1/24` are connected and traffic can be natively routed
+* If source and destination IP belongs to `10.20.0.1/24`, route natively
+* Else, preserve current behavior (encapsulate if being routed to know CIDRs/pass to Linux stack).
+
+### Control plane implementation
+
+Similar to dynamic flowlogs or Hubble dynamic exporter, allow users to configure the CIDRs on the fly through a configmap. The configmap will be mounted on each agent. On create/update, update `LocalNodeConfig` and trigger dataplane reconcile.
+
+### Datapath implementation
+
+This section provides an overview of the datapath implementation. Consider the current implementation - as long as nodes have IP connectivity, Cilium can route traffic, both intra-cluster and inter-cluster (in case of cluster mesh). 
+
+#### Case 1
+
+If `hybrid` mode is enabled without configuring any `subnet-topology`, routing mode will behave exactly same as encapsulation mode. Pod IPs can be from overlay or from a subnet, it will be encapsulated either way. So datapath implementation would be exactly as it is today for encapsulation routing mode.
```

Comment:
Just on wording for clarity:
```suggestion
If `hybrid` mode is enabled without configuring any `subnet-topology`, it will behave the same as `tunnel` mode. Pod IPs can be from the overlay or from a subnet, traffic will be encapsulated either way. So datapath implementation would be exactly as it is today for the `tunnel` routing mode.
```